helix-lsp/src/lib.rs

```diff
@@ -605,109 +599,118 @@ impl Notification {
         }
     }
 
     pub fn get_by_id(&self, id: usize) -> Option<&Client> {
         self.inner
             .values()
-            .find(|(client_id, _)| client_id == &id)
-            .map(|(_, client)| client.as_ref())
+            .find(|client| client.id() == id)
+            .map(|client| &**client)
+    }
+
+    pub fn get_by_name(&self, name: &str) -> Option<&Client> {
+        self.inner.get(name).map(|client| &**client)
     }
 
     pub fn remove_by_id(&mut self, id: usize) {
-        self.inner.retain(|_, (client_id, _)| client_id != &id)
+        self.inner.retain(|_, ls| id != ls.id());
     }
 
-    pub fn restart(
+    fn start_client(
         &mut self,
-        language_config: &LanguageConfiguration,
+        name: String,
+        roots: &[String],
         doc_path: Option<&std::path::PathBuf>,
-    ) -> Result<Option<Arc<Client>>> {
-        let config = match &language_config.language_server {
-            Some(config) => config,
-            None => return Ok(None),
-        };
-
-        let scope = language_config.scope.clone();
-
-        match self.inner.entry(scope) {
-            Entry::Vacant(_) => Ok(None),
-            Entry::Occupied(mut entry) => {
-                // initialize a new client
-                let id = self.counter.fetch_add(1, Ordering::Relaxed);
+    ) -> Result<Arc<Client>> {
+        let config = self
+            .syn_loader
+            .language_server_configs()
+            .get(&name)
+            .ok_or_else(|| anyhow::anyhow!("Language server '{name}' not defined"))?;
+        // initialize a new client
+        // let id = self.counter.fetch_add(1, Ordering::Relaxed);
+        self.counter += 1;
+        let id = self.counter;
+        let NewClient(client, incoming) = start_client(id, name, config, roots, doc_path)?;
+        self.incoming.push(UnboundedReceiverStream::new(incoming));
+        Ok(client)
+    }
 
-                let NewClientResult(client, incoming) =
-                    start_client(id, language_config, config, doc_path)?;
-                self.incoming.push(UnboundedReceiverStream::new(incoming));
+    pub fn restart(
+        &mut self,
+        lang_config: &LanguageConfiguration,
+        doc_path: Option<&std::path::PathBuf>,
+    ) -> Result<Vec<Arc<Client>>> {
+        lang_config
+            .language_servers
+            .iter()
+            .filter_map(|config| {
+                let name = config.name().clone();
+
+                #[allow(clippy::map_entry)]
+                if self.inner.contains_key(&name) {
+                    let client = match self.start_client(name.clone(), &lang_config.roots, doc_path)
+                    {
+                        Ok(client) => client,
+                        error => return Some(error),
+                    };
 
-                let (_, old_client) = entry.insert((id, client.clone()));
+                    let old_client = self.inner.insert(name, client.clone()).unwrap();
```

Comment:
Why not keep the old entry API so we don't have to do a second check when `self.inner.insert`?